### DISCUSSION TITLE
F16 x S8/S4 GEMM

### DISCUSSION THREAD

**Discussion by hwu36** [NVIDIA/cutlass#911] [General]
FastTransformer has kernels written in CUTLASS to support `fp16 x int8/int4` GEMM.

source code:

https://github.com/NVIDIA/FasterTransformer/tree/main/src/fastertransformer/cutlass_extensions/include/cutlass_extensions 

instantiation:

https://github.com/NVIDIA/FasterTransformer/tree/main/src/fastertransformer/kernels/cutlass_kernels/fpA_intB_gemm

paper:

https://arxiv.org/abs/2211.10017

GTC'23 talk:

https://register.nvidia.com/flow/nvidia/gtcspring2023/attendeeportal/page/sessioncatalog/session/1666226207768001N4Fe

**Comment by MARD1NO** [comment]
WeightOnly is very useful in Decoder-only architecture!!!!, but I wonder why use this interleaved layout, I have not seen in cutlass existed layout. 

Is there exist optimizations? I have tried to use the best config in cublasLt int8 TN layout, it seems it's not suitable for interleaved layout...

**Reply by rhenry-nv** [reply]
Hi @MARD1NO,

There are many layout conversions.

There is a reordering of the columns so that we can load contiguous 128 byte cache lines. The GEMM has a K tile size of 64 and the B matrix is int8 and column major. The K-tile means we would normally be reading 64 bytes from a cache line during each load. As a result, I put [64 elements from col1, 64 elements from col2] in 1 cache line to improve the memory access pattern since we will now use all the data in the 128 cache line when it is fetched.

The "interleaving", by this I mean permutation of the rows, was done so that the int8 data ends up in the pattern we need from HMMA after we do LDSM. If we did not interleave the weights this way, we would have to have some communication among the threads to ensure each thread had the "correct" data from the weight matrix before dequantizing and issuing HMMA.

Lastly, there is some permutation within the registers - for int8 is not not needed but it is done so int8 and int4 have the same layout. It is needed for int4 to reduce shift instructions during the conversions. There are also some other transforms done to accelerate the int to float type conversion as @hwu36 pointed out.

The layout is "new" and is different from existing layouts in CUTLASS / cuBLAS since it needed to accomplish different goals.

Hope this clears some things up!

**Reply by MARD1NO** [reply]
Thanks! It really helps me a lot to understand this layout.

**Comment by hwu36** [comment]
I think reordering is used to improve the performance of type conversion https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/cutlass_extensions/include/cutlass_extensions/interleaved_numeric_conversion.h#L54

@byshiue

**Comment by masahi** [comment]
Is this kernel supposed to work on CUDA 12.1? I'm getting strange results where all outputs are zero, even though I confirmed that the kernel is running and inputs are sane. 

On  the container `nvcr.io/nvidia/pytorch:22.09-py3` with CUDA 11.8 that FT supports, I verified that it works correctly.

**Reply by hwu36** [reply]
check ptx and sass between 12.1 and 11.8.  i think 12.1 should work.

**Reply by rhenry-nv** [reply]
What hardware / container / cutlass version are you using? I can't think of a reason 12.1 would break.

**Reply by masahi** [reply]
I'm on RTX 4080 and I want to get the FT kernel working outside of the container. I extracted the FT kernel into https://github.com/masahi/cutlass_fpA_intB_gemm to make it easier to build, and I'm trying to integrate it into TVM. The cutlass version is the same one used by FT (registered as a submodule).

I'm testing my FT kernel + TVM integration. With the debug log enabled (`FT_LOG_LEVEL=TRACE`) I get expected logs from FT like this:
```
[FT][DEBUG] static void fastertransformer::dispatch_stages<T, WeightType, cutlass::arch::Sm80, EpilogueTag, ThreadblockShape, WarpShape, Stages, typename std::enable_if<(Stages > 2)>::type>::dispatch(const T*, const WeightType*, const T*, const T*, T*, int, int, int, fastertransformer::CutlassGemmConfig, char*, size_t, cudaStream_t, int*) [with T = cutlass::half_t; WeightType = cutlass::integer_subbyte<4, false>; EpilogueTag = fastertransformer::EpilogueOpNoBias; ThreadblockShape = cutlass::gemm::GemmShape<128, 128, 64>; WarpShape = cutlass::gemm::GemmShape<128, 32, 64>; int Stages = 4; size_t = long unsigned int; cudaStream_t = CUstream_st*]
[FT][DEBUG] void fastertransformer::generic_mixed_gemm_kernelLauncher(const T*, const WeightType*, const T*, const T*, T*, int, int, int, fastertransformer::CutlassGemmConfig, char*, size_t, cudaStream_t, int*) [with T = cutlass::half_t; WeightType = cutlass::integer_subbyte<4, false>; arch = cutlass::arch::Sm80; EpilogueTag = fastertransformer::EpilogueOpNoBias; ThreadblockShape = cutlass::gemm::GemmShape<128, 128, 64>; WarpShape = cutlass::gemm::GemmShape<128, 32, 64>; int Stages = 4; size_t = long unsigned int; cudaStream_t = CUstream_st*]
[FT][DEBUG] fastertransformer::CutlassFpAIntBGemmRunner<T, WeightType>::~CutlassFpAIntBGemmRunner() [with T = cutlass::half_t; WeightType = cutlass::integer_subbyte<4, false>]
``` 
And I confirmed that the kernel is indeed running without error via (1) adding a print after the kernel launch and (2) nsys:
```
Time (%)  Total Time (ns)  Instances  Avg (ns)  Med (ns)  Min (ns)  Max (ns)  StdDev (ns)                                                  Name                                                
 --------  ---------------  ---------  --------  --------  --------  --------  -----------  ----------------------------------------------------------------------------------------------------
    100.0              800          1     800.0     800.0       800       800          0.0  void cutlass::Kernel<cutlass::gemm::kernel::GemmFpAIntB<cutlass::gemm::threadblock::DqMmaMultistage…
```

Apparently the result matrix is filled with whatever values I initially filled it with. I also tried adding print in the epilogue but it is not showing anything. So maybe there is something very basic off, looking.

**Reply by rhenry-nv** [reply]
I am not sure why, but what you are describing sounds like the kernel is not being compiled due to the macro [here](https://github.com/masahi/cutlass_fpA_intB_gemm/blob/master/cutlass_extensions/include/cutlass_extensions/gemm/kernel/fpA_intB_gemm.h#L479). Can you do something like:
`KernelRunner<true>::run_kernel(params, shared_storage);`

to enable it unconditionally. It was written this way to try to improve compilation speed when doing multiple architectures. And if I understand correctly, you get correct values with 11.8, but nothing runs with 12.1?

**Reply by masahi** [reply]
Turned out I'm hitting this #ifdef for some reason, so the kernel hasn't been compiled at all... 

https://github.com/masahi/cutlass_fpA_intB_gemm/blob/master/cutlass_extensions/include/cutlass_extensions/gemm/kernel/fpA_intB_gemm.h#L483

**Reply by masahi** [reply]
ah yes just saw your message :)

**Reply by rhenry-nv** [reply]
I would also check your cmake file to make sure that we compile for SM89 (SM80 should also be fine). Not sure why else you would fall out of this macro on 4080.

**Reply by rhenry-nv** [reply]
you can check compile commands with verbose make to be sure

**Reply by masahi** [reply]
Apparently `__CUDA_ARCH__` is not defined. Probably something is off in my cmake.

**Reply by masahi** [reply]
Updated my cmake https://github.com/masahi/cutlass_fpA_intB_gemm/commit/bb60c795508798d78a4d7308998db3e034164b52 closely following the one in FT and now I'm getting correct results from FT. Thank you for your help!

**Reply by masahi** [reply]
I've also got vicuna 7b from https://github.com/mlc-ai/mlc-llm running with the FT kernel and I'm getting good-looking tokens generated. Currently I'm not fusing bias or activation but it will be done next week.

I'll continue cleaning up and adding more features to https://github.com/masahi/cutlass_fpA_intB_gemm (support residual fusion, more quantization scheme etc). More models like RedPajama will be supported as well.

**Reply by rhenry-nv** [reply]
Sounds great! Feel free to reach out if you have any more questions.

**Comment by masahi** [comment]
Just found that the FT kernel always uses fp32 accumulation https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/cutlass_extensions/include/cutlass_extensions/gemm/kernel/default_fpA_intB_traits.h#L110

Is there a danger in enabling fp16 accum, at least optionally? `mlc-llm` uses fp16 accum exclusively for vicuna / llama, for example. But for Redpajama it selectively uses fp32 accum.

**Reply by rhenry-nv** [reply]
No danger in having fp16 accumulators [unless your application overflows]. You can add a trait for FP16 accumulators if it helps your application.

**Reply by masahi** [reply]
I tried fp16 accum but the perf didn't change at all. Does that sound weird? I can take a look deeper next week.

The workload is a decoder network (vicuna 7B). So mostly GEMV. Maybe things are so memory bound that the accum dtype in compute doesn't matter?

**Reply by hwu36** [reply]
accumulation type does not impact mma throughput.

**Reply by masahi** [reply]
oh? Then I've been confused about something very basic... 

When people talk about FP16 tensorcore peak flops with fp32 accumulation like below, do they also imply that the output type is fp32? And `ElementOutput` in cutlass does impact mma throughput?

<img width="870" alt="スクリーンショット 2023-06-10 3 04 52" src="https://github.com/NVIDIA/cutlass/assets/1776403/635ed0a1-ed14-441b-a07d-713dea74d946">

**Reply by hwu36** [reply]
let me rephrase.  accumulation type has nothing to do with throughput on a100.  on other cards such as geforce cards, they may do have 2x difference.

sorry, i use a100 every day.  when you say ampere, it means a100 to me. :)

**Comment by masahi** [comment]
I've extracted the FT kernel into https://github.com/tlc-pack/cutlass_fpA_intB_gemm to make it easier to integrate into third-party projects. And I've already made the first improvement over the original implementation: Add support for residual block fusion https://github.com/tlc-pack/cutlass_fpA_intB_gemm/pull/1. 

Things are not documented at all and there is no test either, but nonetheless I hope it would be useful for others as well.

**Reply by hwu36** [reply]
feel free to upstream it to cutlass as an example just like what you did for wgrad and transposed conv.

**Comment by masahi** [comment]
It's being integrated into TVM by https://github.com/apache/tvm/pull/15111

**Comment by manishucsd** [comment]
Hi, All, I am working on making changes to upstream mixed-input support into upstream NVIDIA/CUTLASS. Please review some drawings below on how I am planning to choreograph the mainloop with mixed input datatype. It is slightly different from approach discussed here. 

Notably, we would like to maintain canonical layout in the global memory for TN mixed input (F16 * S8) GEMM and use `__shfl_sync()` after conversion to obtain the data in desired layout for `mma.sync.16816.f32.f16.f16.f32`. I plan to use the conversion tricks discussed here :).

![MixedInput_GEMM_F16xS8](https://github.com/NVIDIA/cutlass/assets/3487276/a9e3e902-82d5-43dd-9be9-1a5314ee3a41)

Consider a warp-level test that I am currently fleshing out using the figure in mind. Let me know if you see an issue in the above approach?

```cpp
TEST(SM80_warp_gemm_tensor_op_mixed_input_crosswise_f16_i8, 64x64x64_64x64x64_16x8x16) {
  using Shape = cutlass::gemm::GemmShape<64, 64, 64>;
  using InstructionShape = cutlass::gemm::GemmShape<16, 8, 16>;
  using ElementA = cutlass::half_t;
  using ElementB = int8_t;
  using ElementC = float;
  using LayoutA = cutlass::layout::RowMajorTensorOpMultiplicandCrosswise<
      cutlass::sizeof_bits<ElementA>::value, 64>;
  using LayoutB = cutlass::layout::ColumnMajorTensorOpMultiplicandCrosswise<
      cutlass::sizeof_bits<ElementB>::value, 64>;

  using MmaTensorOp = typename cutlass::gemm::warp::DefaultMmaTensorOp<
      Shape, InstructionShape, ElementA, LayoutA, ElementB, LayoutB, ElementC,
      cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAddMixedInput>::Type;

  test::gemm::warp::TransformTestbed<MmaTensorOp,
                            cutlass::gemm::GemmShape<64, 64, 64> >()
      .run(cutlass::Distribution::Identity, cutlass::Distribution::Sequential);
}
```
cc: @IonThruster , @hwu36 , @thakkarV , @kerrmudgeon

**Reply by masahi** [reply]
Are you considering dequantization (not cast) for the B matrix?

**Reply by manishucsd** [reply]
Yes, for my first version to check if it all works I might use cast/I2F, but it has to dequantization (as per the tricks discussed here). However, I am looking to use canonical layouts in the global memory for TN GEMM using the figure above.

**Reply by masahi** [reply]
I think the very specific layout used by the FasterTransformer kernel is motivated by the need to make dequantize as fast as possible (sec 3.4 of the paper). ~So the choice of layout and whether you dequantize or not might not be an orthogonal problem.~

UPDATE: just realized that the only part of a dequantization process that affects the layout is I2F.

**Reply by rhenry-nv** [reply]
Yep, exactly @masahi. Only a part of the layout affects I2F. In fact, for int8 we can use the canonical layout and just add a bias to the int8s to use the conversion trick in the paper. 

@manishucsd if you don't want to add a bias to the int8s, you can do it with signed int8s but you need to insert two extra lop3's inside the dequantize sequence. I have it somewhere so I can explain if it is helpful. 

About the proposal
The shuffle seems like a nightmare to get right but the high level of the idea makes a lot of sense to me. One question - are you going to change the threadblock K-tile dimension to 128? 

Also, can I ask why the need to have the canonical layout in global?

**Reply by manishucsd** [reply]
> One question - are you going to change the threadblock K-tile dimension to 128?

I am planning to use and support `ThreadblockShape::kK = 64 and 32`. Thus, it will be 64B and not 128B cache line for S8 (operandB), and I think your questions to address that. Let us see how much of the performance we will lose because of loading partial cachelines. Do you have an estimate? I see Ampere Integer TensorCore GEMMs using `mma.sync.16832.s32.s8.s8.s32` [uses `ThreadblockShape::kK = 128 and 64` ](https://github.com/manishucsd/cutlass/blob/main/tools/library/scripts/generator.py#L2150).

> Also, can I ask why the need to have the canonical layout in global?

Data is already in canonical format, we will need epilogue of the previous operation to write it out in the interleaved layout, if we want to use that for our models. 

Another performance question, can I run the current `F16 * S8` stand alone and measure performance on different GEMM shapes? If yes, can you provide the steps to do it?

**Reply by masahi** [reply]
>  can I run the current F16 * S8 stand alone and measure performance on different GEMM shapes?

Let me plug the work I did to make the FT kernel easier to use (with more features): https://github.com/tlc-pack/cutlass_fpA_intB_gemm. Probably this is the easiest way using only public code. 

You just need to call this function. The weight shape needs to be `(K, N / 2)` for int4, and the scale shape is `(N,)`. 
https://github.com/tlc-pack/cutlass_fpA_intB_gemm/blob/main/cutlass_kernels/fpA_intB_gemm.h#L23-L27

**Reply by rhenry-nv** [reply]
> Let us see how much of the performance we will lose because of loading partial cachelines. Do you have an estimate?

I saw ~10% on some problem sizes where the int8 B matrix was much larger than the activations. This is what motivated the reordering of B to allow the use of 128B cache lines.

> Data is already in canonical format, we will need epilogue of the previous operation to write it out in the interleaved layout, if we want to use that for our models.

Does that mean you will want to use the I2F trick for unsigned int8s?

> Another performance question, can I run the current F16 * S8 stand alone and measure performance on different GEMM shapes? If yes, can you provide the steps to do it?

@masahi 's fork might be the easiest way to use those GEMMs as a stand-alone at the moment.

**Reply by manishucsd** [reply]
I see the following performance with Threadblock::kK = 128 and 64 for `s32 <= s8 * s8 + s32` on A100. 

```
./tools/profiler/cutlass_profiler --m=3456 --n=4096 --k=2048 --output=s8_gemm.csv
=============================
  Problem ID: 1

        Provider: CUTLASS
   OperationKind: gemm
       Operation: cutlass_tensorop_i16832gemm_s8_128x128_64x5_tn_align16

            Math: 450431 GFLOP/s

=============================
  Problem ID: 1

        Provider: CUTLASS
   OperationKind: gemm
       Operation: cutlass_tensorop_i16832gemm_s8_128x128_128x4_tn_align16

            Math: 395882 GFLOP/s
```

> I saw ~10% on some problem sizes where the int8 B matrix was much larger than the activations.

Do you have the problem shapes where you saw that perf difference with 128 vs.  64  elements in tile-K?

@masahi do you have a the changes/pointers ready on profiling the `NVIDIA/FasterTransformer` version of `F32 <= F16 * S8 + F32`?

FYI, I am working on an implementation for mixed input types on canonical layouts with `shfls` to get the mma layout. I will keep you posted, if I could make shfl implementation to work functionally, and then we can look into its performance.

**Reply by masahi** [reply]
> do you have a the changes/pointers ready on profiling the NVIDIA/FasterTransformer version of F32 <= F16 * S8 + F32?

I don't have such code other than the unit tests I added to TVM, which are unnecessarily complex for your purpose.

**Reply by rhenry-nv** [reply]
> Do you have the problem shapes where you saw that perf difference with 128 vs. 64 elements in tile-K?

I think it was M <= 40, N=4096, K=1024, Groups=32. That benchmark was on grouped GEMM, and for very small M which was the initial target for the core MMA. I retrofitted it to regular GEMM later.

**Reply by dongjiyingdjy** [reply]
Hi，these weight only quantized GEMM helps a lot in LLMs. However, the problem size of GEMM is larger. The most commonly used shape is M <=64, N=4096, K=16384. 
Have you tried K tile size more than 64, for example, 128 or 256? Could a bigger size of K tile improve memory bandwidth for large K?

**Comment by tridao** [comment]
My understanding is that the I2F conversion is done after loading from smem to rmem, right before calling `mma.sync`. Is that right? And the loading from smem to rmem is still using `ldmatrix`?

Did you guys ever try doing I2F conversion right after loading from gmem, before writing to smem? That might make the code simpler, but idk how much it affects performance. Even though that would increase the amount of smem read/write, maybe this kind of gemm is bottlenecked by either gmem bandwidth or fp16 mma anyway, and not smem read/write?

**Reply by manishucsd** [reply]
> My understanding is that the I2F conversion is done after loading from smem to rmem, right before calling mma.sync. Is that right? And the loading from smem to rmem is still using ldmatrix?

yes and yes

> Did you guys ever try doing I2F conversion right after loading from gmem, before writing to smem? 

Gmem to Smem is done using cp.async which doesn't bring registers in between. cp.async reads from Gmem and writes data directly into Smem.

**Reply by rhenry-nv** [reply]
+1 to above. There is support for what @tridao described in the [Mmapipeline ](https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/cutlass_extensions/include/cutlass_extensions/gemm/threadblock/dq_mma_pipelined.h#L89 )version of those kernels. You can control apply a transform either after LDG or after ldmatrix. However, we don't use that main loop on Ampere.

If I remember correctly, Volta does the transform after LDG while Turing does it after ldmatrix.

**Comment by alexsamardzic** [comment]
Any specific plans to actually have any of this upstream in CUTLASS soon?  I was not aware of anything above, and as mentioned [here](https://github.com/NVIDIA/FasterTransformer/issues/492), in the meantime I was working on [utilizing it in PyTorch](https://github.com/pytorch/pytorch/pull/107782), starting from CUTLASS extensions from the FasterTransformer project (mine is f16xs8 only at the moment, but at least CUTLASS extensions from FasterTransformer are updated for CUTLASS 3.x in my version).  Of course, it would be much better to actually have this functionality in CUTLASS.

**Reply by manishucsd** [reply]
Your timing couldn't be any better for this request. I am on it and very close to starting a PR to NVIDIA/CUTLASS.

**Comment by masahi** [comment]
Hi @rhenry-nv, I'm wondering if it is possible to run the FT int4/8 GEMM kernel on multiple GPUs. 

The way https://github.com/mlc-ai/mlc-llm does multi-gpu for non-FT paths is to shard the quantized weight along row or column dimension, do GEMM on each device, and do NCCL AllReduce to gather the results. This scheme doesn't seem to work for the FT kernel due to the need for weight preprocessing, involving elements permute and transpose. Any thought?

**Reply by MARD1NO** [reply]
I think you can preprocess weight + split weight first, then you can run on multiple gpus.

**Reply by masahi** [reply]
The result is incorrect if I do that.

**Reply by rhenry-nv** [reply]
Yes it it possible. You must split first then preprocess. The contiguous dimension of the matrix must be a multiple of 64 after splitting.

**Reply by masahi** [reply]
hmm ok. I thought the transposition of the weight would pose a problem for such independent preprocessing of sharded weights, but I'll try.

**Comment by MARD1NO** [comment]
oh sry，I made wrong. the correct order is split weight first, then do preprocess on each device's weight.&nbsp;



---Original---
From: ***@***.***&gt;
Date: Wed, Oct 18, 2023 21:36 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [NVIDIA/cutlass] F16 x S8/S4 GEMM (Discussion #911)




 
The result is incorrect if I do that.

**Reply by masahi** [reply]
I thought about that, but preprocessing involves transpose of the entire matrix. Split and transpose would end up with a different matrix.

**Comment by fishelegs** [comment]
Will cutlass support group quantization for S4, in F16 x S4?
Since in LLM 4bit PTQ will lose some ppl, 4 bit group quantization will improve ppl a lot.

**Reply by manishucsd** [reply]
Hi @fishelegs , 

Thanks for bringing this to our attention. We are looking into and going to look look more closely into it in the coming weeks. Stay tuned. 

Can you please expand your acronyms here?
PTQ: Post Training Quantization
PPL?

**Reply by fishelegs** [reply]
1）PPL stands for perplexity
Perplexity (PPL) is one of the most common metrics for evaluating language models.
(ref:https://huggingface.co/docs/transformers/perplexity)

2）I find that TensorRT-LLM seems to have the 4bit group-wise weight-only quantization.
(https://github.com/NVIDIA/TensorRT-LLM/blob/d8ebeee2f6fcb219e6efc541ccc914765799fa3a/examples/llama/README.md?plain=1#L104)
In Section "Groupwise quantization (AWQ/GPTQ)"
I gonna have a try.

**Comment by MARD1NO** [comment]
@rhenry-nv Hi rawn henry，I found in cutlass official's mixed input gemm using ColMajor weight layout, and use FragementShuffler to get the mma layout's weight. **Is there exist performance gap between ColMajor and InterleavedLayout?** I think InterleavedLayout may be get better performance, but it is hard to expand to other device and dtype. Using Manish's mixed input gemm, it is easier to expand like fp8 matmul int4.

**Reply by manishucsd** [reply]
**FP**: FP8 should not need a new implementation `FragmentShuffler`, but for FP8 which is on Hopper you can use `ld.shared` instead of `ldmatrix` with similar performance avoiding the need for `FragmentShuffler`. These should be in CUTLASS 3.4 and @rhenry-nv can confirm. 

**S4**: S4 will need a new implementation of `FragmentShuffler` and @alexsamardzic was working on it. 

I agree using canonical ColMajor layout for quantized weights makes it easier to exploit a heterogenous compute offering as Interleaved layout are device-specific.

**Comment by sleepwalker2017** [comment]
> FastTransformer has kernels written in CUTLASS to support `fp16 x int8/int4` GEMM.
> 
> source code:
> 
> https://github.com/NVIDIA/FasterTransformer/tree/main/src/fastertransformer/cutlass_extensions/include/cutlass_extensions
> 
> instantiation:
> 
> https://github.com/NVIDIA/FasterTransformer/tree/main/src/fastertransformer/kernels/cutlass_kernels/fpA_intB_gemm
> 
> paper:
> 
> https://arxiv.org/abs/2211.10017
> 
> GTC'23 talk:
> 
> https://register.nvidia.com/flow/nvidia/gtcspring2023/attendeeportal/page/sessioncatalog/session/1666226207768001N4Fe

Hi  what is the name of the GTC talk? this link is too old and is lost. 

Could you provide the name of this talk?  Thank you so much !

**Reply by sleepwalker2017** [reply]
Is it this talk? https://www.nvidia.com/en-us/on-demand/session/gtcspring23-s51226/
